pkg/destroy/aws/aws.go

```diff
@@ -1994,3 +2058,54 @@ func removeSharedTag(ctx context.Context, tagClients []*resourcegroupstaggingapi
 
 	return nil
 }
+
+// destroyedResources stores the results of attempts to destroy resources. The key is the ARN of the resource. The value
+// is whether the resource was successfully destroyed.
+type destroyedResources map[string]bool
+
+// record records where a resource was successfully destroyed. If the specified error is nil, then the resource was
+// destroyed successfully.
+func (r destroyedResources) record(arn string, err error) {
+	r[arn] = (err == nil)
+}
+
+// merge merges the record of destroyed resources from changes into r. If changes and r have results for the same
+// resource, the results from changes have precedence.
+func (r destroyedResources) merge(changes destroyedResources) {
+	for k, v := range changes {
+		r[k] = v
+	}
+}
+
+// extractRemaining removes the resources that could not be successfully deleted from r and returns a slice containing
+// the ARNs of those resources.
+func (r destroyedResources) extractRemaining() []string {
+	if r == nil {
+		return nil
```

Comment:
it feels strange to me to have a `nil` guard here but not in `merge` or `record`.  Is there a reason behind the difference?

Reply:
A nil guard in `merge` or `record` would mean that the operation did not do anything. The functions cannot maps to merge or record into that will be accessible to the caller.

This works the same way as the operations on a general map. You can get from a nil map, but you cannot set into a nil map.

Reply:
If you feel better about it, I can get rid of the nil guard. The function will return an empty slice instead of a nil slice, in that case.

Reply:
> A nil guard in `merge` or `record` would mean that the operation did not do anything.

Makes sense.  Where are we initializing these maps today?  Do we need a:

```go
func newDestroyedResources() destroyedResources {
  return map[string]bool{}
}
```

or something?

Reply:
They are initialized with `destroyedResources{}`. I don't think we need to introduce a function to do that.